File: C++/maximum-depth-of-binary-tree.cpp
#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

void maxDepthHelper(TreeNode *node, int&depth, int currDepth);

int maxDepth(TreeNode* root) {
    int depth = 0;
    int currDepth = 1;
    maxDepthHelper(root, depth, currDepth);
    return depth;
}

void maxDepthHelper(TreeNode *node, int &depth, int currDepth) {
    if (node == nullptr) return;
    if (currDepth > depth) {
        depth = currDepth;
    }
    maxDepthHelper(node -> left, depth, currDepth + 1);
    maxDepthHelper(node -> right, depth, currDepth + 1);
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
	
	// insert driver code
	
	return 0;
}

File: Python/maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getMaximumDepth(self, root, currentDepth, maxDepth):
        if root == None:
            return 0

        newDepth = currentDepth + 1

        leftMaxDepth = self.getMaximumDepth(root.left, newDepth, maxDepth)
        rightMaxDepth = self.getMaximumDepth(root.right, newDepth, maxDepth)

        maxDepth = max(maxDepth, newDepth, leftMaxDepth, rightMaxDepth)
        return maxDepth

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        return self.getMaximumDepth(root, 0, 0)